File: src/fixengine/messages/Fields.java
/*
 * Copyright 2009 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package fixengine.messages;

import java.nio.ByteBuffer;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.Map;

import fixengine.tags.MsgType;

/**
 * @author Pekka Enberg
 */
public class Fields implements Iterable<Field> {
    private final Map<Integer, Field> fields = new LinkedHashMap<Integer, Field>();

    @Override public Iterator<Field> iterator() {
        return fields.values().iterator();
    }

    public Field lookup(Tag<?> tag) {
        return fields.get(tag.value());
    }

    public Field lookup(int tag) {
        return fields.get(tag);
    }

    public void parse(ByteBuffer b) {
        Tag<?> previous = MsgType.TAG;
        for (;;) {
            b.mark();
            int tag = Tag.parseTag(b, previous);
            Field field = lookup(tag);
            if (field == null)
                break;
            previous = field.parse(b);
            if (!field.isFormatValid())
                throw new InvalidValueFormatException(field.prettyName() + ": Invalid value format");
            if (!field.isValueValid())
                throw new InvalidValueException(field.prettyName() + ": Invalid value");
        }
        b.reset();
    }

    public String format() {
        StringBuilder result = new StringBuilder();
        for (Field field : fields.values()) {
            result.append(field.format());
        }
        return result.toString();
    }

    public void validate() {
        for (Field field : fields.values()) {
            if (field.isEmpty())
                throw new EmptyTagException(field.prettyName() + ": Empty tag");
        }
    }

    public void add(Tag<?> tag, Required required) {
        this.fields.put(tag.value(), tag.newField(required));
    }
}

File: src/main/scala/xtch/itch/elements/Fields.scala
/*
 * Copyright 2010 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package xtch.itch.elements

import xtch.itch.types.{Alpha, Numeric}

object Fields {
  val Attribution = Field("Attribution", Alpha(4))
  val BuyOrSellIndicator = Field("BuyOrSellIndicator", Alpha(1))
  val EventCode = Field("EventCode", Alpha(1))
  val FinancialProduct = Field("FinancialProduct", Numeric(3))
  val ISIN = Field("ISIN", Alpha(12))
  val MIC = Field("MIC", Alpha(4))
  val MarketSegmentID = Field("MarketSegmentID", Numeric(3))
  val Millisecond = Field("Millisecond", Numeric(3))
  val NoteCodes = Field("NoteCodes", Numeric(8))
  val OrderBook = Field("OrderBook", Numeric(6))
  val OrderReferenceNumber = Field("OrderReferenceNumber", Numeric(9))
  val Price = Field("Price", Numeric(10))
  val Quantity = Field("Quantity", Numeric(9))
  val Reason = Field("Reason", Alpha(4))
  val Reserved = Field("Reserved", Alpha(1))
  val RoundLotSize = Field("RoundLotSize", Numeric(9))
  val Second = Field("Second", Numeric(5))
  val Symbol = Field("Symbol", Alpha(16))
  val TradingCurrency = Field("TradingCurrency", Alpha(3))
  val TradingState = Field("TradingState", Alpha(1))
}
